TestFoundation/TestNSLock.swift

```diff
@@ -0,0 +1,152 @@
+            Thread.sleep(forTimeInterval: 8)
+            lock.unlock()
+            self.debugLog("Released lock")
+        }
+        condition.lock()
+        thread.start()
+        condition.wait()
+        condition.unlock()
+        debugLog("Thread started")
+
+        XCTAssertTrue(test(waitTime: 0, shouldLock: false))
+        XCTAssertTrue(test(waitTime: -1, shouldLock: false))
+        XCTAssertTrue(test(waitTime: 1, shouldLock: false))
+        XCTAssertTrue(test(waitTime: 4, shouldLock: false))
+        XCTAssertTrue(test(waitTime: 8, shouldLock: true))
+        XCTAssertTrue(test(waitTime: -1, shouldLock: true))
+    }
+
+
+    func test_threadsAndLocks() {
+        let condition = NSCondition()
+        let lock = NSLock()
+        let threadCount = 1000
+        let endSeconds: Double = 10
+
+        let endTime = Date.init(timeIntervalSinceNow: endSeconds)
+        var threadsStarted = Array<Bool>(repeating: false, count: threadCount)
+        let arrayLock = NSLock()
+
+        for t in 0..<threadCount {
+            let thread = Thread() {
+                arrayLock.lock()
+                threadsStarted[t] = true
+                arrayLock.unlock()
+
+                let name = Thread.current.name ?? "unknown"
+
+                condition.lock()
+                condition.wait()
+                condition.unlock()
+                for _ in 1...100 {
+                    let r = (endSeconds * drand48()) / 100
+                    Thread.sleep(forTimeInterval: r)
+                    if lock.lock(before: endTime) {
+                        self.debugLog("\(name): got lock")
+                        lock.unlock()
+                    }
+                }
+                arrayLock.lock()
+                threadsStarted[t] = false
+                arrayLock.unlock()
+            }
+            thread.name = "Thread \(t)"
+            thread.start()
+        }
+
+        var totalThreads = 0
+        repeat {
+            arrayLock.lock()
+            totalThreads = threadsStarted.filter {$0 == true }.count
+            arrayLock.unlock()
+            debugLog("Have \(totalThreads) started")
+        } while totalThreads < threadCount
+        XCTAssertEqual(totalThreads, threadCount)
+
+        debugLog("Broadcasting to \(threadCount) threads")
+        condition.lock()
+        condition.broadcast()
+        condition.unlock()
+
+        debugLog("Initial start")
+        Thread.sleep(until: endTime)
+        debugLog("Waiting for threads to finish")
+        repeat {
+            arrayLock.lock()
+            totalThreads = threadsStarted.filter {$0 == false }.count
+            arrayLock.unlock()
+            debugLog("Have \(totalThreads) finished")
+        } while totalThreads < threadCount
+        XCTAssertEqual(totalThreads, threadCount)
+
+        let gotLock = lock.try()
+        XCTAssertTrue(gotLock)
+        if gotLock {
+            lock.unlock()
+        }
+    }
+}
```

Comment:
Out of curiosity, how long do these tests take to run? Is there going to be a problem with creating so many threads on some platforms?

Reply:
I just ran the tests on my 2014 Macbook Pro and they took 8.008 and 10.003 seconds so about 18seconds for both tests.

1000 threads were created in .15secs. I was originally creating 10,000 but the limit looks to be 2048 total (on macOS)  and I could only create 2044. It didn't error out but I think `Thread.start` could be modified to get the error result from `pthread_create` and set the status appropriately.

I guess some low end devices will have a problem but the `threadCount` can just be lowered, the idea is really to try and stress the system to force some locking errors.

I have some more tests to add at some point for the `NSConditionLock` so I'll tidy up the debug then.